controllers/infra/zone/zone_controller_test.go

```diff
@@ -0,0 +1,307 @@
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha3"
+	"github.com/vmware-tanzu/vm-operator/controllers/infra/zone"
+	topologyv1 "github.com/vmware-tanzu/vm-operator/external/tanzu-topology/api/v1alpha1"
+	pkgcfg "github.com/vmware-tanzu/vm-operator/pkg/config"
+	"github.com/vmware-tanzu/vm-operator/pkg/constants/testlabels"
+	pkgctx "github.com/vmware-tanzu/vm-operator/pkg/context"
+	providerfake "github.com/vmware-tanzu/vm-operator/pkg/providers/fake"
+	"github.com/vmware-tanzu/vm-operator/pkg/util/kube/cource"
+	"github.com/vmware-tanzu/vm-operator/pkg/util/ptr"
+	vsclient "github.com/vmware-tanzu/vm-operator/pkg/util/vsphere/client"
+	"github.com/vmware-tanzu/vm-operator/pkg/util/vsphere/watcher"
+	vmwatcher "github.com/vmware-tanzu/vm-operator/services/vm-watcher"
+	"github.com/vmware-tanzu/vm-operator/test/builder"
+	"github.com/vmware-tanzu/vm-operator/test/testutil"
+)
+
+var _ = Describe(
+	"Reconcile",
+	Label(
+		testlabels.Controller,
+		testlabels.EnvTest,
+		testlabels.V1Alpha3,
+	),
+	func() {
+
+		var (
+			ctx       context.Context
+			vcSimCtx  *builder.IntegrationTestContextForVCSim
+			provider  *providerfake.VMProvider
+			initEnvFn builder.InitVCSimEnvFn
+		)
+
+		BeforeEach(func() {
+			ctx = context.Background()
+			ctx = logr.NewContext(ctx, testutil.GinkgoLogr(4))
+		})
+
+		JustBeforeEach(func() {
+			ctx = pkgcfg.WithContext(ctx, pkgcfg.Default())
+			ctx = pkgcfg.UpdateContext(
+				ctx,
+				func(config *pkgcfg.Config) {
+					config.Features.WorkloadDomainIsolation = true
+				},
+			)
+			ctx = cource.WithContext(ctx)
+			ctx = watcher.WithContext(ctx)
+
+			provider = providerfake.NewVMProvider()
+			provider.VSphereClientFn = func(ctx context.Context) (*vsclient.Client, error) {
+				return vsclient.NewClient(ctx, vcSimCtx.VCClientConfig)
+			}
+
+			vcSimCtx = builder.NewIntegrationTestContextForVCSim(
+				ctx,
+				builder.VCSimTestConfig{
+					WithWorkloadIsolation: pkgcfg.FromContext(ctx).Features.WorkloadDomainIsolation,
+				},
+				func(ctx *pkgctx.ControllerManagerContext, mgr ctrlmgr.Manager) error {
+					if err := vmwatcher.AddToManager(ctx, mgr); err != nil {
+						return err
+					}
+					return zone.AddToManager(ctx, mgr)
+				},
+				func(ctx *pkgctx.ControllerManagerContext, _ ctrlmgr.Manager) error {
+					ctx.VMProvider = provider
+					return nil
+				},
+				initEnvFn)
+			Expect(vcSimCtx).ToNot(BeNil())
+
+			vcSimCtx.BeforeEach()
+
+			ctx = vcSimCtx
+		})
+
+		AfterEach(func() {
+			vcSimCtx.AfterEach()
+		})
+
+		When("new zones are added", func() {
+			var (
+				nsInfo builder.WorkloadNamespaceInfo
+			)
+
+			JustBeforeEach(func() {
+				nsInfo = vcSimCtx.CreateWorkloadNamespace()
+
+				By("ensure all zones have finalizers", func() {
+					Eventually(func(g Gomega) {
+						var obj topologyv1.ZoneList
+						g.Expect(vcSimCtx.Client.List(ctx, &obj, ctrlclient.InNamespace(nsInfo.Namespace))).To(Succeed())
+						g.Expect(obj.Items).ToNot(BeEmpty())
```

Comment:
Also assert the number of expected zones - `g.Expect(obj.Items).To(HaveLen(vcSimCtx.))`